zerver/actions/message_send.py

```diff
@@ -1525,19 +1525,63 @@ def validate_stream_id_with_pm_notification(
     return stream
 
 
-def check_private_message_policy(
-    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile]
+def check_can_send_direct_message(
+    realm: Realm, sender: UserProfile, user_profiles: Sequence[UserProfile], recipient: Recipient
 ) -> None:
-    if realm.private_message_policy == PrivateMessagePolicyEnum.DISABLED:
-        if sender.is_bot or (
-            len(user_profiles) == 1 and (user_profiles[0].is_bot or user_profiles[0] == sender)
-        ):
-            # We allow direct messages only between users and bots or to oneself,
-            # to avoid breaking the tutorial as well as automated
-            # notifications from system bots to users.
-            return
+    if sender.is_bot:
+        return
+
+    if all(user_profile.is_bot or user_profile.id == sender.id for user_profile in user_profiles):
+        return
 
-        raise JsonableError(_("Direct messages are disabled in this organization."))
+    direct_message_permission_group = realm.direct_message_permission_group
+
+    # TODO: This check is inefficient; we should in the future be able to cache
+    # on the Huddle object whether the conversation already exists, likely in the
+    # form of a `first_message_id` field, and be able to save doing this check in the
+    # common case that this is not the first message in a conversation.
+    if (
+        not hasattr(direct_message_permission_group, "named_user_group")
+        or direct_message_permission_group.named_user_group.name != SystemGroups.EVERYONE
+    ):
+        allowed_user_ids = get_user_group_member_ids(direct_message_permission_group)
+        if sender.id not in allowed_user_ids:
+            for user_profile in user_profiles:
+                if user_profile.id in allowed_user_ids:
+                    break
+            else:
+                raise JsonableError(
+                    _("You are not allowed to send direct messages to these recipient(s).")
+                )
+
+    direct_message_initiator_group = realm.direct_message_initiator_group
+    if (
+        not hasattr(direct_message_initiator_group, "named_user_group")
+        or direct_message_initiator_group.named_user_group.name != SystemGroups.EVERYONE
+    ):
+        if is_user_in_group(direct_message_initiator_group, sender):  #  nocoverage
+            return
+        if recipient.type == Recipient.PERSONAL:
+            recipient_user_profile = user_profiles[0]
+            previous_messages_exist = (
+                Message.objects.filter(
+                    realm=realm,
+                    recipient__type=Recipient.PERSONAL,
+                )
+                .filter(
+                    Q(sender=sender, recipient__type_id=recipient_user_profile.id)
+                    | Q(sender=recipient_user_profile, recipient__type_id=sender.id)
```

Comment:
These checks are weird, and only work because of the `recipient__type` check above. Just filter on `recipient=recipient_user_profile.recipient`, or `sender.recipient`; we have those fields denormalized.

Reply:
I have changed `recipient__type_id=sender.id`  to `recipient=sender.recipient` but using `recipient=recipient_user_profile.recipient` requires one extra query. Are we OK with that?

Reply:
`recipient_id=recipient_user_profile.recipient_id` should avoid the query.